workflow: functionality changes release notes addition

0xc0170 opened the pull request:
Every functionality change must contain release notes text addition. This will be
gating functionality change integration.

A user does not need to go through PRs to see what has changed, read the section:
"Functionality changes" in the release notes.

Ready for reviews

cc @ARMmbed/mbed-os-maintainers 

We could get some examples from current funcitonal changes and how release notes should look like. I quickly found one: https://github.com/ARMmbed/mbed-os/pull/9277

0xc0170 committed:
workflow: functionality changes release notes addition

Every functionality change must contain release notes text addition. This will be
gating functionality change integration.

A user does not need to go through PRs to see what has changed, read the section:
"Functionality changes" in the release notes.

NirSonnenschein commented on docs/reference/contributing/guidelines/workflow.md:
```diff
@@ -98,6 +98,12 @@ Any change in the functionality, it can be adding a new feature, adding a new me
 
 A feature contribution contains a new API, capability or behavior. It does not break backward compatibility with existing APIs, capabilities or behaviors. New feature contributions are very welcome in Mbed OS. However, because they add capability to the codebase, it's easy for a new feature to introduce bugs and a support burden. The introduction of new features should also come with documentatio ...
 
+Every functionality change pull request must contain release notes section addition to describe the change to users.
```
must contain release notes section -> must contain a release notes section
I would also propose a title for this section 
so it would be:
must contain a release notes section called "Important changes" 

"important changes" is a placeholder actual name still needs to be defined

0xc0170 replied:
fixed, adding "Release notes" section for now

NirSonnenschein commented on docs/reference/contributing/guidelines/workflow.md:
```diff
@@ -98,6 +98,12 @@ Any change in the functionality, it can be adding a new feature, adding a new me
+Every functionality change pull request must contain release notes section addition to describe the change to users.
+
+It must contain:
+- description of changes with impact analysis 
```
you might want to give examples for "description of changes" and "impact analysis" as developers can interpret these in different ways.

0xc0170 replied:
I split them into two points

NirSonnenschein reviewed (commented):
look promising

0xc0170 committed:
workflow: fix impact analysis vs description

NirSonnenschein commented on docs/reference/contributing/guidelines/workflow.md:
```diff
@@ -98,6 +98,13 @@ Any change in the functionality, it can be adding a new feature, adding a new me
+It must contain:
+- brief description of changes introduced
+- impact analysis - identify components affected, what are potential consequences for users, why do we need this
+- migration guidance: before and after (good to include code snippets to illustrate)
```
small note: in migration guidance I suggest saying it needs to specify all the actions that the user needs to take (should be self explanatory, but still).

0xc0170 replied:
+1, I rephrased it

NirSonnenschein reviewed (approved):
approved, with one small note

0xc0170 committed:
workflow: functionality change migration fix

0xc0170 commented on docs/reference/contributing/guidelines/workflow.md:
```diff
@@ -98,6 +98,13 @@ Any change in the functionality, it can be adding a new feature, adding a new me
+- brief description of changes introduced
+- impact analysis - identify components affected, what are potential consequences for users, why do we need this
+- migration guidance: actions for updating the current code. Good to include code snippets to illustrate, before/after
+
```
Maybe add a note here t hat this section will be used in official release notes ? or it's clear to reader

adbridge replied:
Agreed, a note is useful

bulislaw commented on docs/reference/contributing/guidelines/workflow.md:
```diff
@@ -98,6 +98,13 @@ Any change in the functionality, it can be adding a new feature, adding a new me
 
 A feature contribution contains a new API, capability or behavior. It does not break backward compatibility with existing APIs, capabilities or behaviors. New feature contributions are very welcome in Mbed OS. However, because they add capability to the codebase, it's easy for a new feature to introduce bugs and a support burden. The introduction of new features should also come with documentatio ...
 
+Every functionality change pull request must contain a release notes section called "Release notes" to describe the changes to users.
```
`Every pull request changing functionality`? @AnotherButler

0xc0170 replied:
fixed in efc2e0b

adbridge replied:
Should also apply for 'Breaking Changes'

bulislaw commented on docs/reference/contributing/guidelines/workflow.md:
```diff
@@ -98,6 +98,13 @@ Any change in the functionality, it can be adding a new feature, adding a new me
 
 A feature contribution contains a new API, capability or behavior. It does not break backward compatibility with existing APIs, capabilities or behaviors. New feature contributions are very welcome in Mbed OS. However, because they add capability to the codebase, it's easy for a new feature to introduce bugs and a support burden. The introduction of new features should also come with documentatio ...
 
+Every functionality change pull request must contain a release notes section called "Release notes" to describe the changes to users.
```
I'm not sure if this paragraph keeps the spirit of the previous ones. The first and second paragraph talks exclusively about adding new functionality. Your paragraph below talks more about changing existing functionality.

0xc0170 replied:
I rephrased it to "Every pull request changing or adding functionality" - we want to capture additions but also changes

bulislaw reviewed (commented):
Should we rewrite the "Functionality change" to talk both about new functionality but also about changing functionality. The whole page should be reviewed by docs team btw.

0xc0170 committed:
workflow: functionality change or addition fix

0xc0170 commented:
Example for https://github.com/ARMmbed/mbed-os/pull/9277 (@deepikabhavnani can help me to fix wording here - will edit).

### Release notes

Add `MBED_TRAP_ERRORS_ENABLED` to develop profile - enables to catch printf in ISR in develop profile (it was only enabled in debug profile previously). These errors were silent in develop profile. This change might result in the final binary size increase.

0xc0170 commented:
@AnotherButler This is very important change to be included in the process as soon as possible. Please review

deepikabhavnani commented:
> Example for ARMmbed/mbed-os#9277 (@deepikabhavnani can help me to fix wording here - will edit).

Corrected a bit, but still @AnotherButler  should verify 

Release notes
Added `MBED_TRAP_ERRORS_ENABLED` flag to develop profile - Traps on RTX errors and invalid operations in ISR context enabled in develop profile. These errors were silent in develop profile.
This change might result in the final binary size increase.

Commit:
Edit workflow.md

Edit file, mostly for complete sentences and parallelism.

AnotherButler commented on docs/reference/contributing/guidelines/workflow.md:
```diff
@@ -94,10 +94,18 @@ Release: patch
+It must contain:
+
+- A brief description of changes introduced.
+- An impact analysis: components affected, potential consequences for users and reasons for the addition or change.
```
Query: Can we use a word other than "impact" here? It's technically imprecise language. ("Impact" implies hitting something, such as impacted teeth or a meteor impact.)

0xc0170 replied:
OK, what other synonym could be used? I've checked some, could not find anything similar to this one.

I found usually "impact analysis" in some software journals. See for instance https://www.guru99.com/impact-analysis-software-testing.html

0xc0170 committed:
workflow: impact to effects wording fix

0xc0170 commented:
@AnotherButler Fixed impact as proposed, I would still go with as it was. `An analysis of effects` works as well

adbridge commented:
> 
> 
> Uhm, I like the idea of less manual work, but it also means we need to be validating it during code review. It would be good to have our release notes grammatically correct and describes the change properly (with analysis of impact and migration steps when it makes sense).

For functionality changes and breaking changes where a 'Release notes' we should add docs team as reviewers just for that section.

Commit:
Add requested note to workflow.md

Add requested note to clarify content will be used in the official release note.

AnotherButler commented:
Do we know why Travis is failing?

0xc0170 commented:
Travis having networking issues, restarted the build

0xc0170 commented:
@AnotherButler All green, lets merge this in and let us know once it's in, we will bookmark this in the docs to share with the teams
